packages/playground/blueprints/src/lib/zip.ts

```diff
@@ -0,0 +1,753 @@
+		end: number
+	) => Promise<ReadableStream<Uint8Array>>;
+};
+
+async function createFetchSource(
+	url: string,
+	contentLength?: number
+): Promise<BytesSource> {
+	if (contentLength === undefined) {
+		contentLength = await getContentLength(url);
+	}
+
+	return {
+		length: contentLength,
+		streamBytes: async (from: number, to: number) =>
+			await fetch(url, {
+				headers: {
+					Range: `bytes=${from}-${to}`,
+					'Accept-Encoding': 'none',
+				},
+			}).then((response) => response.body!),
+	};
+}
+
+function concatUint8Array(...arrays: Uint8Array[]) {
+	const result = new Uint8Array(
+		arrays.reduce((sum, array) => sum + array.length, 0)
+	);
+	let offset = 0;
+	for (const array of arrays) {
+		result.set(array, offset);
+		offset += array.length;
+	}
+	return result;
+}
+
+// @ts-ignore
+if (!ReadableStream.prototype[Symbol.asyncIterator]) {
+	// @ts-ignore
+	ReadableStream.prototype[Symbol.asyncIterator] = async function* () {
+		const reader = this.getReader();
+		try {
+			while (true) {
+				const { done, value } = await reader.read();
+				if (done) {
+					return;
+				}
+				yield value;
+			}
+		} finally {
+			reader.releaseLock();
+		}
+	};
+}
+
+type IterableReadableStream<R> = ReadableStream<R> & {
+	[Symbol.asyncIterator](): AsyncIterableIterator<R>;
+} & AsyncIterable<R>;
+
+function filterStream<T>(filter: (chunk: T) => boolean) {
+	return new TransformStream<T, T>({
+		transform(chunk, controller) {
+			if (filter(chunk)) {
+				controller.enqueue(chunk);
+			}
+		},
+	});
+}
+
+const DEFAULT_PREDICATE = () => true;
+export async function iterateRemoteZip(
+	url: string,
+	predicate: (
+		dirEntry: CentralDirectoryEntry | FileEntry
+	) => boolean = DEFAULT_PREDICATE
+) {
+	if (predicate === DEFAULT_PREDICATE) {
+		// If we're not filtering the zip contents, let's just
+		// grab the entire zip.
+		const response = await fetch(url);
+		return iterateZipStream(response.body!);
+	}
+
+	const contentLength = await getContentLength(url);
+	if (contentLength <= PREFER_RANGES_IF_FILE_LARGER_THAN) {
+		// If the zip is small enough, let's just grab it.
+		const response = await fetch(url);
+		return iterateZipStream(response.body!);
+	}
+
+	// Ensure ranges query support:
+	// Fetch one byte
+	const response = await fetch(url, {
+		headers: {
+			Range: 'bytes=0-0',
+			'Accept-Encoding': 'none',
+		},
+	});
+
+	// Fork the stream so that we can reuse it in case
+	// bytes are unsupported and we're now streaming the
+	// entire file
+	const [peekStream, responseStream] = response.body!.tee();
+
+	// Read from the forked stream and confirm we only
+	// got a single byte.
+	const peekReader = peekStream.getReader();
+	const { value: peekBytes } = await peekReader.read();
+	const { done: peekDone } = await peekReader.read();
+	// Close the forked stream
+	peekReader.releaseLock();
+	peekStream.cancel();
+
+	// Confirm our Range query worked as intended:
+	const rangesSupported = peekBytes?.length === 1 && peekDone;
```

Comment:
why is `peekBytes.length` 1 when we ask for `0-0`?

Reply:
I was confused myself! The values are inclusive, so the range starts at byte `0` and ends after byte `0` – which is just the byte `0`. That's a good opportunity to leave a comment.

Reply:
I updated it to say `The Range header is inclusive, so we need to subtract 1`.